tsdb/shard.go

```diff
@@ -66,49 +69,38 @@ type Shard struct {
 	LogOutput io.Writer
 }
 
-// ShardConfig is passed to NewShard to specify the shard's
-// database, retention policy, and location of files on disk.
-type ShardConfig struct {
-	// Name of the database this shard belongs to
-	Database string
-
-	// Name of the retention policy this shard belongs to
-	RetentionPolicy string
-
-	// Path to this shard's location on disk
-	Path string
-
-	// Path to this shard's WAL location
-	WALPath string
-}
-
 // NewShard returns a new initialized Shard. walPath doesn't apply to the b1 type index
-func NewShard(id uint64, index *DatabaseIndex, config ShardConfig, options EngineOptions) *Shard {
+func NewShard(id uint64, index *DatabaseIndex, path string, walPath string, options EngineOptions) *Shard {
 	// Configure statistics collection.
-	key := fmt.Sprintf("shard:%s:%d", config.Path, id)
+	key := fmt.Sprintf("shard:%s:%d", path, id)
+	db, rp := DecodeStorePath(path)
 	tags := map[string]string{
-		"path":            config.Path,
+		"path":            path,
 		"id":              fmt.Sprintf("%d", id),
 		"engine":          options.EngineVersion,
-		"database":        config.Database,
-		"retentionPolicy": config.RetentionPolicy,
+		"database":        db,
+		"retentionPolicy": rp,
 	}
 	statMap := influxdb.NewStatistics(key, "shard", tags)
 
 	return &Shard{
 		index:             index,
 		id:                id,
-		config:            config,
+		path:              path,
+		walPath:           walPath,
 		options:           options,
 		measurementFields: make(map[string]*MeasurementFields),
 
+		database:        db,
+		retentionPolicy: rp,
+
 		statMap:   statMap,
 		LogOutput: os.Stderr,
 	}
 }
 
 // Path returns the path set on the shard when it was created.
-func (s *Shard) Path() string { return s.config.Path }
+func (s *Shard) Path() string { return s.path }
```

Comment:
While this doesn't _need_ a read lock on `s.mu`, I think it should either have one, or the one in `DiskSize` should be removed. There are no concurrent writes to `s.path`, so technically no lock is currently needed, but it would be better to be consistent either way..

If it's decided to add a read lock to `Path()` then you should also move `path` under `mu` in the `Shard` type definition.